server/api/impl/admin/synchronize.go

```diff
@@ -0,0 +1,225 @@
+			values := map[string]string{}
+			if _, ok := updatedAppVersionsMap[registeredApp.Manifest.AppID]; ok {
+				values[oldVersionKey] = updatedAppVersionsMap[registeredApp.Manifest.AppID]
+			}
+			// Call onStartup the function of the app. It should be called only once
+			f := func() error {
+				if err := adm.call(registeredApp, registeredApp.Manifest.OnStartup, values); err != nil {
+					adm.mm.Log.Error("Can't call onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+				}
+				return nil
+			}
+			if err := adm.callOnce(f); err != nil {
+				adm.mm.Log.Error("Can't callOnce the onStartup func of the app", "app_id", registeredApp.Manifest.AppID, "err", err.Error())
+			}
+		}
+	}
+
+	return nil
+}
+
+func (adm *Admin) DeleteApp(app *apps.App) error {
+	// Call delete the function of the app
+	if err := adm.call(app, app.Manifest.OnDelete, nil); err != nil {
+		return errors.Wrapf(err, "delete failed. appID - %s", app.Manifest.AppID)
+	}
+
+	// delete oauth app
+	conf := adm.conf.GetConfig()
+	client := model.NewAPIv4Client(conf.MattermostSiteURL)
+	client.SetToken(string(adm.adminToken))
+
+	if app.OAuth2ClientID != "" {
+		success, response := client.DeleteOAuthApp(app.OAuth2ClientID)
+		if !success || response.StatusCode != http.StatusNoContent {
+			return errors.Wrapf(response.Error, "failed to delete OAuth2 App - %s", app.Manifest.AppID)
+		}
+	}
+
+	// delete app from proxy plugin, not removing the data
+	if err := adm.store.DeleteApp(app); err != nil {
+		return errors.Wrapf(err, "can't delete app - %s", app.Manifest.AppID)
+	}
+
+	adm.mm.Log.Info("Deleted the app", "app_id", app.Manifest.AppID)
+
+	return nil
+}
+
+func (adm *Admin) AddApp(manifest *apps.Manifest) error {
+	newApp := &apps.App{}
+	newApp.Manifest = manifest
+	newApp.Status = apps.AppStatusRegistered
+	if err := adm.store.StoreApp(newApp); err != nil {
+		return errors.Wrapf(err, "can't store app - %s", manifest.AppID)
+	}
+	adm.mm.Log.Info("App is registered", "app_id", manifest.AppID)
+	return nil
+}
+
+func (adm *Admin) callOnce(f func() error) error {
+	// Delete previous job
+	key := "PP_CallOnce_key"
+	if err := adm.mm.KV.Delete(key); err != nil {
+		return errors.Wrap(err, "can't delete key")
+	}
+	// Ensure all instances run this
+	time.Sleep(10 * time.Second)
+
+	adm.mutex.Lock()
+	defer adm.mutex.Unlock()
+	value := 0
+	if err := adm.mm.KV.Get(key, &value); err != nil {
+		return err
+	}
+	if value != 0 {
+		// job is already run by other instance
+		return nil
+	}
+
+	// job is should be run by this instance
+	if err := f(); err != nil {
+		return errors.Wrap(err, "can't run the job")
+	}
+	value = 1
+	ok, err := adm.mm.KV.Set(key, value)
+	if err != nil {
+		return errors.Wrapf(err, "can't set key %s to %d", key, value)
+	}
+	if !ok {
+		return errors.Errorf("can't set key %s to %d", key, value)
+	}
+	return nil
+}
+
+func (adm *Admin) call(app *apps.App, call *apps.Call, values map[string]string) error {
```

Comment:
1/5 `call` is not very intention-revealing. 0/5 Perhaps `expandedAdminCall`? Seems a handful to type/read. `adminCall`? Do all `Call`s made by Admin interface go through this function? If not, the name should reflect the intention.